airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py

```diff
@@ -164,27 +167,28 @@ def is_done(self) -> bool:
         return (
             not self._streams_currently_generating_partitions
             and not self._stream_instances_to_start_partition_generation
-            and all([all(p.is_closed() for p in partitions) for partitions in self._streams_to_partitions.values()])
+            and all([len(running_partitions) == 0 for running_partitions in self._streams_to_running_partitions.values()])
+            and len(self._streams_done) == len(self._stream_name_to_instance)
         )
 
     def _is_stream_done(self, stream_name: str) -> bool:
-        return (
-            all([p.is_closed() for p in self._streams_to_partitions[stream_name]])
-            and stream_name not in self._streams_currently_generating_partitions
+        return stream_name in self._streams_done or (
+            len(self._streams_to_running_partitions[stream_name]) == 0 and stream_name not in self._streams_currently_generating_partitions
```

Comment:
I don't like the condition `len(self._streams_to_running_partitions[stream_name]) == 0 and stream_name not in self._streams_currently_generating_partitions`

This means that the following assertion is valid even though we haven't processed anything:
```
        handler = ConcurrentReadProcessor(...)
        assert handler._is_stream_done(_STREAM_NAME)
```

For me, it either means that the name is not right or the implementation is not right. It feels like it's the later but this would be harder to fix as we would need to check if the stream has had any partition being processed

Reply:
This jumped out at me too. It feels like there should be a more direct way to tell if the stream is done, e.g. by confirming that the sentinels have been seen.